lib/puma/minissl.rb

```diff
@@ -54,22 +54,21 @@ def read_nonblock(size, *_)
           output = engine_read_all
           return output if output
 
-          begin
-            data = @socket.read_nonblock(size, exception: false)
-            if data == :wait_readable || data == :wait_writable
-              if @socket.to_io.respond_to?(data)
-                @socket.to_io.__send__(data)
-              elsif data == :wait_readable
-                IO.select([@socket.to_io])
-              else
-                IO.select(nil, [@socket.to_io])
-              end
-            elsif !data
-              return nil
-            else
-              break
-            end
-          end while true
+          data = @socket.read_nonblock(size, exception: false)
+          if data == :wait_readable || data == :wait_writable
+            # It would make more sense to let @socket.read_nonblock raise
+            # EAGAIN if necessary but it seems like it'll misbehave on Windows.
+            # I don't have a Windows machine to debug this so I can't explain
+            # exactly whats happening in that OS. Please let me know if you
+            # find out!
```

Comment:
@MSP-Greg can you verify/clarify this comment from @Jesus, so we can remove it based on your feedback?

Reply:
I changed:
```ruby
@socket.read_nonblock(size, exception: false)
```
to
```ruby
@socket.read_nonblock(size)
```
and made the other adjustments.

On Windows every job failed, but Travis passed.  See:
https://ci.appveyor.com/project/MSP-Greg/puma/build/job/ykqnya8f3y1ehorf
https://travis-ci.org/MSP-Greg/puma/builds/567079580

I'll investigate further when I have more time, it may have something to do with the method returning nil, but I'm not sure.

Regardless, this seems to be a Windows quirk.

Reply:
JFYI, I quickly checked Ruby's OpenSSL tests, and also WEBrick.  Anywhere that `read_nonblock` could raise an error, `exception: false` was used.  Interesting thing is that many of the commits were done before testing was done on Windows...